# بسم الله الرحمن الرحيم

In [1]:
# import pandas as pd 
# import numpy as np 
# from matplotlib import pyplot as plt 
# import seaborn as sns 
# plt.style.use('default')
# #-------------------------------------------------------------------------------------
# from sklearn.utils import resample 
# from sklearn import preprocessing
# from sklearn.linear_model import LogisticRegression
# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.svm import SVC
# from sklearn.pipeline import make_pipeline
# from sklearn.metrics import confusion_matrix
# from sklearn.model_selection import cross_validate

# from sklearn.naive_bayes import GaussianNB
# from sklearn.metrics import confusion_matrix
# from sklearn.metrics import classification_report
# from sklearn.metrics import precision_score
# from sklearn.metrics import recall_score
# from sklearn.pipeline import Pipeline 

# import tensorflow_addons as tfa
# import tensorflow as tf 
# from tensorflow import keras 
# from tensorflow.keras import layers 
# from focal_loss import SparseCategoricalFocalLoss

# #-------------------------------------------------------------------------------------

# from keras.callbacks import EarlyStopping
# from keras.callbacks import ModelCheckpoint
# #-------------------------------------------------------------------------------------
# from imblearn.over_sampling import SMOTE
# from utils import *
# %matplotlib inline

In [2]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split


2023-05-13 08:00:37.804872: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-13 08:00:37.985710: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-05-13 08:00:37.985730: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-05-13 08:00:38.890623: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2023-

## Reading the Dataset, cleaning , preparing it for the training 

In [3]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [4]:
df_original= cleaning_data(df).copy()

In [5]:
all_class_train,all_class_test=data_spliting(df_original,1/3)

In [6]:
X_train=all_class_train.drop(["Body_Level"],axis=1).to_numpy()
Y_train=all_class_train["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [7]:
X_train, scaler= data_scaling(X_train)


In [8]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [9]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [10]:
model=build_model(3,0.01)

2023-05-13 08:00:40.441420: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-05-13 08:00:40.441470: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2023-05-13 08:00:40.441519: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (brownboyComputer): /proc/driver/nvidia/version does not exist
2023-05-13 08:00:40.441954: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [11]:
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=300 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/before_sampling.h5")])

Epoch 1/300
31/31 [==============================] - 1s 8ms/step - loss: 0.6514 - accuracy: 0.5112 - val_loss: 0.3507 - val_accuracy: 0.5633
Epoch 2/300
31/31 [==============================] - 0s 2ms/step - loss: 0.3159 - accuracy: 0.6379 - val_loss: 0.2580 - val_accuracy: 0.6653
Epoch 3/300
31/31 [==============================] - 0s 3ms/step - loss: 0.2331 - accuracy: 0.7140 - val_loss: 0.2181 - val_accuracy: 0.7714
Epoch 4/300
31/31 [==============================] - 0s 2ms/step - loss: 0.1921 - accuracy: 0.7647 - val_loss: 0.1967 - val_accuracy: 0.7918
Epoch 5/300
31/31 [==============================] - 0s 3ms/step - loss: 0.1683 - accuracy: 0.7870 - val_loss: 0.1828 - val_accuracy: 0.8041
Epoch 6/300
31/31 [==============================] - 0s 3ms/step - loss: 0.1528 - accuracy: 0.8012 - val_loss: 0.1696 - val_accuracy: 0.8204
Epoch 7/300
28/31 [==========================>...] - ETA: 0s - loss: 0.1444 - accuracy: 0.8181
Yes You are here 0.820486843585968 0.8244897723197937
31/31

31/31 [==============================] - 0s 2ms/step - loss: 0.0434 - accuracy: 0.9351 - val_loss: 0.0544 - val_accuracy: 0.9469
Epoch 43/300
 1/31 [..............................] - ETA: 0s - loss: 0.0456 - accuracy: 0.9688
Yes You are here 0.9391480684280396 0.9387755393981934
31/31 [==============================] - 0s 3ms/step - loss: 0.0420 - accuracy: 0.9391 - val_loss: 0.0524 - val_accuracy: 0.9388
Epoch 44/300
 1/31 [..............................] - ETA: 0s - loss: 0.0680 - accuracy: 0.8125
Yes You are here 0.9350912570953369 0.9428571462631226
31/31 [==============================] - 0s 3ms/step - loss: 0.0409 - accuracy: 0.9351 - val_loss: 0.0525 - val_accuracy: 0.9429
Epoch 45/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0408 - accuracy: 0.9361 - val_loss: 0.0520 - val_accuracy: 0.9306
Epoch 46/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0394 - accuracy: 0.9422 - val_loss: 0.0511 - val_accuracy: 0.9388
Epoch 47/300
31/31 [=======

31/31 [==============================] - 0s 3ms/step - loss: 0.0215 - accuracy: 0.9675 - val_loss: 0.0304 - val_accuracy: 0.9796
Epoch 90/300
 1/31 [..............................] - ETA: 0s - loss: 0.0123 - accuracy: 1.0000
Yes You are here 0.9736308455467224 0.9795918464660645
31/31 [==============================] - 0s 3ms/step - loss: 0.0213 - accuracy: 0.9736 - val_loss: 0.0289 - val_accuracy: 0.9796
Epoch 91/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0209 - accuracy: 0.9675 - val_loss: 0.0294 - val_accuracy: 0.9796
Epoch 92/300
 1/31 [..............................] - ETA: 0s - loss: 0.0242 - accuracy: 0.9688
Yes You are here 0.970588207244873 0.9795918464660645
31/31 [==============================] - 0s 3ms/step - loss: 0.0205 - accuracy: 0.9706 - val_loss: 0.0289 - val_accuracy: 0.9796
Epoch 93/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0202 - accuracy: 0.9706 - val_loss: 0.0289 - val_accuracy: 0.9755
Epoch 94/300
31/31 [========

Epoch 136/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0140 - accuracy: 0.9777 - val_loss: 0.0222 - val_accuracy: 0.9796
Epoch 137/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0142 - accuracy: 0.9716 - val_loss: 0.0227 - val_accuracy: 0.9796
Epoch 138/300
 1/31 [..............................] - ETA: 0s - loss: 0.0138 - accuracy: 1.0000
Yes You are here 0.9746450185775757 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0139 - accuracy: 0.9746 - val_loss: 0.0224 - val_accuracy: 0.9837
Epoch 139/300
 1/31 [..............................] - ETA: 0s - loss: 0.0110 - accuracy: 1.0000
Yes You are here 0.9776875972747803 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0138 - accuracy: 0.9777 - val_loss: 0.0220 - val_accuracy: 0.9837
Epoch 140/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0140 - accuracy: 0.9767 - val_loss: 0.0221 - val_accuracy: 0.9796
Epoch 141

31/31 [==============================] - 0s 3ms/step - loss: 0.0110 - accuracy: 0.9807 - val_loss: 0.0194 - val_accuracy: 0.9796
Epoch 175/300
 1/31 [..............................] - ETA: 0s - loss: 0.0060 - accuracy: 1.0000
Yes You are here 0.9776875972747803 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0112 - accuracy: 0.9777 - val_loss: 0.0216 - val_accuracy: 0.9837
Epoch 176/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0111 - accuracy: 0.9797 - val_loss: 0.0193 - val_accuracy: 0.9796
Epoch 177/300
 1/31 [..............................] - ETA: 0s - loss: 0.0083 - accuracy: 1.0000
Yes You are here 0.9787018299102783 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0109 - accuracy: 0.9787 - val_loss: 0.0198 - val_accuracy: 0.9837
Epoch 178/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0105 - accuracy: 0.9838 - val_loss: 0.0207 - val_accuracy: 0.9796
Epoch 179/300
31/31 [==

Epoch 208/300
 1/31 [..............................] - ETA: 0s - loss: 0.0049 - accuracy: 1.0000
Yes You are here 0.9817444086074829 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0095 - accuracy: 0.9817 - val_loss: 0.0182 - val_accuracy: 0.9837
Epoch 209/300
 1/31 [..............................] - ETA: 0s - loss: 0.0060 - accuracy: 1.0000
Yes You are here 0.976673424243927 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0101 - accuracy: 0.9767 - val_loss: 0.0191 - val_accuracy: 0.9837
Epoch 210/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0096 - accuracy: 0.9746 - val_loss: 0.0191 - val_accuracy: 0.9796
Epoch 211/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0095 - accuracy: 0.9797 - val_loss: 0.0190 - val_accuracy: 0.9796
Epoch 212/300
31/31 [==============================] - 0s 2ms/step - loss: 0.0088 - accuracy: 0.9838 - val_loss: 0.0188 - val_accuracy: 0.9796
Epoch 213/

31/31 [==============================] - 0s 4ms/step - loss: 0.0083 - accuracy: 0.9828 - val_loss: 0.0195 - val_accuracy: 0.9796
Epoch 244/300
26/31 [========================>.....] - ETA: 0s - loss: 0.0075 - accuracy: 0.9820
Yes You are here 0.9807302355766296 0.9836734533309937
31/31 [==============================] - 0s 4ms/step - loss: 0.0082 - accuracy: 0.9807 - val_loss: 0.0216 - val_accuracy: 0.9837
Epoch 245/300
25/31 [=======================>......] - ETA: 0s - loss: 0.0073 - accuracy: 0.9887
Yes You are here 0.9858012199401855 0.9836734533309937
31/31 [==============================] - 0s 4ms/step - loss: 0.0078 - accuracy: 0.9858 - val_loss: 0.0203 - val_accuracy: 0.9837
Epoch 246/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0077 - accuracy: 0.9848 - val_loss: 0.0196 - val_accuracy: 0.9796
Epoch 247/300
30/31 [============================>.] - ETA: 0s - loss: 0.0079 - accuracy: 0.9833
Yes You are here 0.982758641242981 0.9836734533309937
31/31 [========

31/31 [==============================] - 0s 3ms/step - loss: 0.0099 - accuracy: 0.9767 - val_loss: 0.0225 - val_accuracy: 0.9796
Epoch 282/300
 1/31 [..............................] - ETA: 0s - loss: 0.0133 - accuracy: 0.9375
Yes You are here 0.9868153929710388 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0071 - accuracy: 0.9868 - val_loss: 0.0198 - val_accuracy: 0.9837
Epoch 283/300
 1/31 [..............................] - ETA: 0s - loss: 0.0033 - accuracy: 1.0000
Yes You are here 0.9847870469093323 0.9836734533309937
31/31 [==============================] - 0s 3ms/step - loss: 0.0071 - accuracy: 0.9848 - val_loss: 0.0222 - val_accuracy: 0.9837
Epoch 284/300
30/31 [============================>.] - ETA: 0s - loss: 0.0076 - accuracy: 0.9854
Yes You are here 0.9858012199401855 0.9836734533309937
31/31 [==============================] - 0s 4ms/step - loss: 0.0076 - accuracy: 0.9858 - val_loss: 0.0230 - val_accuracy: 0.9837
Epoch 285/300
27/31 [=======

In [12]:
accurate_train_accuracy,classify_report= test_model("NNModels/before_sampling.h5",X_train,Y_train)

31/31 [==============================] - 0s 1ms/step


In [13]:
accurate_train_accuracy

0.9929006085192698

In [14]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.98      1.00      0.99       127
           1       0.99      0.96      0.97       134
           2       0.99      1.00      0.99       271
           3       1.00      1.00      1.00       454

    accuracy                           0.99       986
   macro avg       0.99      0.99      0.99       986
weighted avg       0.99      0.99      0.99       986



In [15]:
accurate_val_accuracy,classify_report= test_model("NNModels/before_sampling.h5",X_val,Y_val)

8/8 [==============================] - 0s 952us/step


In [16]:
accurate_val_accuracy

0.9836734693877551

In [17]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        29
           1       1.00      0.97      0.98        31
           2       0.96      0.99      0.97        73
           3       0.99      0.98      0.99       112

    accuracy                           0.98       245
   macro avg       0.99      0.98      0.99       245
weighted avg       0.98      0.98      0.98       245



In [18]:
accurate_test_accuracy,classify_report= test_model("NNModels/before_sampling.h5",X_test,Y_test)

8/8 [==============================] - 0s 1ms/step


In [19]:
accurate_test_accuracy

0.975609756097561

In [20]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.97      0.94      0.96        34
           1       0.94      0.94      0.94        36
           2       0.95      1.00      0.98        62
           3       1.00      0.98      0.99       114

    accuracy                           0.98       246
   macro avg       0.97      0.97      0.97       246
weighted avg       0.98      0.98      0.98       246



## important-Note
gamma=1 on-test(0.967479674796748) on-val(0.9836734693877551)
gamma=2 on-test(0.967479674796748) on-val(0.9836734693877551) - No difference